### ISSUE TITLE
Kestrel: Is it possible for an app to control the lifetime/disposal of each request's IMemoryOwner?

### ISSUE THREAD

**Issue by hiteshmadan** [dotnet/aspnetcore#21385]
### Is your feature request related to a problem? Please describe.
I'm looking for a way to use the byte buffers underlying the kestrel request stream and control when they get disposed. I couldn't find any way to stop kestrel from returning the bytes back to its' pool when the request ends.

### Describe the solution you'd like
A way to tell kestrel to not dispose/not reuse the byte arrays for any request (either on a per request basis; or a server setting). + a way to get the IMemoryOwner that has the request body.

### Additional context
This task is for a message bus that receives events on a HTTP API, queues them in an in-mem channel, and pushes them out to one or more user-configured HTTP endpoint/Service Bus topic /Event Hub namespaces.
(think eventgrid running in a single process, without persistence). Since the incoming request should not wait for the message to be pushed out on the other end, I end up copying the incoming request's bytes to my own pooled byte[] today and this takes a good 7-10% CPU.

**Comment by davidfowl** [comment]
No it isn't, that would require reference counting which System.IO.Pipelines doesn't support. This isn't something we'd support natively as this level of manual memory management is hard to get right (like really really hard). The problem is everyone has to play nice in these systems and it requires advanced diagnostics to debug memory ownership issues (this is what happens when you stop using the GC).

Kestrel *could* allow changing the memory pool implementation which would let you implement your own reference counting semantics. When you own the pool then you own the memory and you can use MemoryMarshel.TryGetMemoryManager to get the backing IMemoryOwner for a Memory and up the reference count, and "take ownership" of the buffer. It's like poor mans rust 😄.

PS: Most of those messaging APIs currently take an array today anyway. You won't be able to use a Memory\<byte\> or even an ArraySegment\<byte\> so you're basically forced to copy.

**Comment by hiteshmadan** [comment]
Changing memory pool could work - except that iirc kestrel's internal memory pool implementation performs much better than .net core's default MemoryPool implementation. I'm guessing it won't help the overall picture since the 7-10% CPU saved on byte[] copying will now be taken up by a relatively worse memory pool implementation.

Would you mind sharing why you see this as devolving into a full blown ref counting solution?
I can see "transfer of ownership" via the IMemoryOwner interface : if someone does a context.Request.GetBodyMemoryOwner(), kestrel has handed over the ownership of the underlying byte[] to the caller and can now skip the "return byte[] to underlying pool" step when the request ends. Potentially guard this API with an opt-in flag in kestrel's settings too, to avoid people accidentally shooting themselves in the foot.

 PS: I am the messaging API here (eventgrid on edge) :)

**Comment by davidfowl** [comment]
> Changing memory pool could work - except that iirc kestrel's internal memory pool implementation performs much better than .net core's default MemoryPool implementation. I'm guessing it won't help the overall picture since the 7-10% CPU saved on byte[] copying will now be taken up by a relatively worse memory pool implementation.

You can copy it. 

> Would you mind sharing why you see this as devolving into a full blown ref counting solution?

You're thinking about using this in a very narrow situation where you're going to immediately going to enqueue that work into a "well behaved" API. We need to think about cases where that won't happen, and you can pass around this memory to any API. There need to be a usable model since we're subverting the type system.

> I can see "transfer of ownership" via the IMemoryOwner interface : if someone does a context.Request.GetBodyMemoryOwner(), kestrel has handed over the ownership of the underlying byte[] to the caller and can now skip the "return byte[] to underlying pool" step when the request ends.

We wouldn't do that by default for a bunch of reasons:
- The ones I listed on this thread. I don't trust anybody to get this right (we experimented with this earlier on and we got it all wrong ourselves). I'm good having them own all of the memory though. That's the price for this level of control.
- We never have the body as a single memory owner, it's potentially split across multiple memory owners.

> PS: I am the messaging API here (eventgrid on edge) :)

Oh good 😄, make the API support `Memory<byte>` and `ReadOnlySequence<byte>`

**Comment by hiteshmadan** [comment]
Thanks for the explanation! Having a way to replace the MemoryPool looks good to me then.

> Oh good 😄, make the API support ReadOnlyMemory<byte> and ReadOnlySequence<byte>

Just for context on the flow as it stands today (and why its not a problem of byte[] vs ROM<byte> vs ROS<byte> from what I can tell) :
```
API:
Kestrel incoming request 
-> Copy context.Request.Body to an IMemoryOwner<byte> rented from the shared Memory Pool
-> Enqueue IMemoryOwner<byte> to an in-memory channel 
-> Respond with http ok

Background workers:
Dequeue IMemoryOwner from channel
-> Http send to destination 
-> IMemoryOwner.Dispose()
```

Its the Copy request body stage that I'm trying to eliminate here.

**Comment by davidfowl** [comment]
> Its the Copy request body stage that I'm trying to eliminate here.

Yep, I'm familiar with the problem and the safety tradeoff isn't worth it IMO.
